File: tree/Construct_all_possible_BSTs_for_keys_1_to_N/construct.js
class Node{
    constructor(data){
        this.data = data; 
        this.right = null ; 
        this.left = null ;
    } 
}

function constructTree(start, end){
    let lst = []
    if(start > end){
        lst.push(null) ; 
        return lst ; 
    }

    for(var i = start; i <= end; i++){
        let leftsubtree = constructTree(start, i-1) ; 
        let rightsubtree = constructTree(i+1, end); 
        for(var j = 0 ; j < leftsubtree.length ; j++) {
            let left = leftsubtree[j] ; 
            for(var k = 0 ; k < rightsubtree.length; k++){
                let right = rightsubtree[k] ; 
                let node = new Node(i) ; 
                node.right = right ; 
                node.left = left ; 
                
                lst.push(node) ; 
            }
        }
    }

    return lst ; 
}

function preorder(node){
    if(node == null){
        return ; 
    }

    console.log(node.data, " ") ; 
    preorder(node.left); 
    preorder(node.right) ; 
}

var list = constructTree(1, 5) 
for (var i = 0 ; i < list.length; i++){
    console.log("Preorder traversal : ") ; 
    preorder(list[i]) ; 
    console.log("\n") ; 
}

File: tree/Construct_all_possible_BSTs_for_keys_1_to_N/construct.cpp
#include <iostream>
#include <vector> 
using namespace std; 

class Node {
    public : 
        int data ; 
        Node* right ; 
        Node* left ; 

        Node(int data){
            this->data = data ; 
            this->left = NULL ; 
            this->right = NULL ; 
        }
} ; 
vector<Node*> constructTree(int start, int end){
    vector<Node*> lst ;
    if(start > end){
        lst.push_back(NULL);

        return lst ; 
    }

    for(int i = start; i <= end; i++){
        vector<Node*> leftsubtree = constructTree(start, i-1) ;
        vector<Node*> rightsubtree = constructTree(i+1, end) ; 
        for(int j = 0 ; j < leftsubtree.size(); j++){
            Node* left = leftsubtree[j] ; 
            for(int k = 0 ; k < rightsubtree.size(); k++){
                Node* right = rightsubtree[k] ; 
                Node* root = new Node(i) ; 
                root->left = left ; 
                root->right = right ; 

                lst.push_back(root);
            }
        } 
    }
    return lst ; 
}

void Preorder(Node* node){
    if(node == NULL){
        return ;
    }
    cout<<node->data<<" " ; 
    Preorder(node->left);
    Preorder(node->right) ; 
}

int main(){
    vector<Node*> lst = constructTree(1, 5);
    for(int i = 0 ; i < lst.size() ; i++){
        cout<<"Preorder traversal : "; 
        Preorder(lst[i]);
        cout<<"\n" ; 
    }

    return 0 ; 
}